src/optimize_1q_gates.rs

```diff
@@ -0,0 +1,102 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2022
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use pyo3::prelude::*;
+use pyo3::wrap_pyfunction;
+use pyo3::Python;
+
+const PI: f64 = std::f64::consts::PI;
+
+//     Return a triple theta, phi, lambda for the product.
+//
+//         u3(theta, phi, lambda)
+//            = u3(theta1, phi1, lambda1).u3(theta2, phi2, lambda2)
+//            = Rz(phi1).Ry(theta1).Rz(lambda1+phi2).Ry(theta2).Rz(lambda2)
+//            = Rz(phi1).Rz(phi').Ry(theta').Rz(lambda').Rz(lambda2)
+//            = u3(theta', phi1 + phi', lambda2 + lambda')
+//
+//         Return theta, phi, lambda.
+#[pyfunction]
```

Comment:
One thing we normally do for `pyfunction` is to also set a text_signature macro so that `inspect` and similar tools (include some IDEs) can reason about the signature of the function from Python. Something like:

```suggestion
#[pyfunction]
#[pyo3(text_signature = "(theta1, phi1, lambda1, theta2, phi2, lambda2, /)")]
```